blockmanager.go

```diff
@@ -1019,109 +1022,153 @@ func (b *blockManager) getCheckpointedCFHeaders(checkpoints []*chainhash.Hash,
 			// they're too far ahead, or discard them if we don't
 			// need them.
 
-			// Find the first and last height for the blocks
-			// represented by this message.
-			startHeight := checkPointIndex*wire.CFCheckptInterval + 1
-			lastHeight := (nextCheckPointIndex + 1) * wire.CFCheckptInterval
+			select {
+			case headerChan <- r:
+			case <-b.quit:
+				return false, false
+			}
 
-			log.Debugf("Got cfheaders from height=%v to "+
-				"height=%v, prev_hash=%v", startHeight,
-				lastHeight, r.PrevFilterHeader)
+			return true, true
+		}
 
-			// If this is out of order but not yet written, we can
-			// verify that the checkpoints match, and then store
-			// them.
-			if startHeight > curHeight+1 {
-				log.Debugf("Got response for headers at "+
-					"height=%v, only at height=%v, stashing",
-					startHeight, curHeight)
+		msgs = append(msgs, &query.Query{
+			Req:        m,
+			HandleResp: handleResp,
+		})
+	}
 
-				queryResponses[checkPointIndex] = r
+	// Hand the queries to the work manager, and consume the verified
+	// responses as they come back.
+	errChan := b.cfg.QueryAccess.Query(msgs)
 
-				return true
-			}
+	// Keep waiting for more headers as long as we haven't received an
+	// answer for our last checkpoint, and no error is encountered.
+	for {
+		var r *wire.MsgCFHeaders
+		select {
+		case r = <-headerChan:
+		case err := <-errChan:
+			log.Errorf("Query finished with error (%v) before "+
+				"all responses received", err)
+			return
+		case <-b.quit:
+			return
+		}
 
-			// If this is out of order stuff that's already been
-			// written, we can ignore it.
-			if lastHeight <= curHeight {
-				log.Debugf("Received out of order reply "+
-					"end_height=%v, already written", lastHeight)
-				return true
-			}
+		checkPointIndex, ok := stopHashes[r.StopHash]
```

Comment:
Nit: no need to check `ok` here as we already did before this.

Reply:
Done